File: WebSite/EulerAList/pset14.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <vector>

using namespace std;

unordered_map<long, long> path;

long findPath(long x)
{
    if (x == 1)
        return 0;
    if (path.find(x) != path.end())
        return path[x];
    vector<long> usedValue;
    while (path.find(x) == path.end())
    {
        usedValue.push_back(x);
        if (x % 2)
        {
            x = 3 * x + 1;
        }
        else
        {
            x /= 2;
        }
    }
    long terminalValue = path[x];
    for (int i = usedValue.size() - 1; i >= 0; i --)
    {
        terminalValue ++;
        path[usedValue[i]] = terminalValue;
    }
    return terminalValue;
}

int main()
{
    long result = 1;
    long maxPath = 1;
    path[1] = 0;
    for (long i = 2; i < 1000000; i ++)
    {
        long x = findPath(i);
        if (maxPath < x)
        {
            maxPath = x;
            result = i;
        }
    }
    cout << result <<endl;
    return 0;
}

File: WebSite/EulerAList/pset14.py
path = {}

def findPath(x):
    if x == 1:
        return 0
    if x in path.keys():
        return path[x]
    usedValue = []
    while not x in path.keys():
        usedValue.append(x)
        if x % 2:
            x = 3 * x + 1
        else:
            x /= 2
    terminalValue = path[x]
    for i in range(len(usedValue) - 1, 0, -1):
        terminalValue += 1
        path[usedValue[i]] = terminalValue 
    return terminalValue

result = 1
maxPath = 1
path[1] = 0
for i in range(2, 1000000, 1):
    x = findPath(i)
    if maxPath < x:
        maxPath = x
        result = i
print(result)
